File: Java/DynamicProgramming/BJAlgo_12865.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.Arrays;

public class BJAlgo_12865 {
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int[] nk = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
        int[][] dp = new int[nk[0] + 1][nk[1] + 1];
        for (int i = 1; i < nk[0] + 1; i++) {
            int[] wv = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
            for (int j = 1; j < nk[1] + 1; j++) {
                if (j < wv[0]) {
                    dp[i][j] = dp[i - 1][j];
                } else {
                    dp[i][j] = Math.max(dp[i - 1][j], dp[i - 1][j - wv[0]] + wv[1]);
                }
            }
        }

        System.out.println(dp[nk[0]][nk[1]]);
    }
}

File: BaekjoonAlgorithm/202012/BA_12865/BJAlgo_12865.py
from sys import stdin

n, k = map(int, stdin.readline().split())
dp = [[0] * (k + 1) for _ in range(n + 1)]
for i in range(1, n + 1):
    w, v = map(int, stdin.readline().split())
    for j in range(1, k + 1):
        if j < w:
            dp[i][j] = dp[i - 1][j]
        else:
            dp[i][j] = max(dp[i - 1][j - w] + v, dp[i - 1][j])

print(dp[-1][-1])
